<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="roomba_core"> <!-- Robot core: for Joints/links -->
<xacro:include filename="inertial_macros.xacro"/> 
<xacro:include filename="properties.xacro"/> 

    <!-- Define reusable materials -->
    <material name="white"> 
        <color rgba="1 1 1 1" />
    </material>

    <material name="orange">
        <color rgba="1 0.3 0.1 1"/>
    </material>

    <material name="blue">
        <color rgba="0.2 0.2 1 1"/>
    </material>

    <material name="black">
        <color rgba="0 0 0 1"/>
    </material>

    <material name="grey">
      <color rgba="0.5 0.5 0.5 1"/>
    </material>

    <material name="transparent">
      <color rgba="0 0 0 0"/>
    </material>

    <material name="red">
      <color rgba="1 0 0 1"/>
    </material>

  <!--=== PROPERTIES ===-->
  <!-- All dimensions in meters -->
  
  <!--=== BASE LINK (CHASSIS AS A CYLINDER) ===-->
  <!-- <origin xyz="0 0 ${chassis_thick/2}" rpy="0 0 0"/> -->
  <link name="base_link">
    <visual>
      <origin xyz="0 0 ${chassis_thick/2}" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="${chassis_radius}" length="${chassis_thick}"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <!-- In gazebo/simulation environment, it should bump/respond in realism terms -->
    <collision>
      <origin xyz="0 0 ${chassis_thick/2}" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="${chassis_radius}" length="${chassis_thick}"/>
      </geometry>
    </collision>
    <!-- mass: 3 -->
    <xacro:inertial_cylinder mass="3.0"
                             length="${chassis_thick}"
                             radius="${chassis_radius}">
      <origin xyz="0 0 ${chassis_thick/2}" rpy="0 0 0"/>
    </xacro:inertial_cylinder>

  </link>
  <gazebo reference="base_link">
    <material>Gazebo/Grey</material>
    <gravity>false</gravity>
  </gazebo>
  <!--=== LEFT WHEEL ===-->
  <!-- base_link then left_wheel_joint then left_wheel -->
  <!-- This deals with position ONLY, relative to parent (need both joint/link to define position of rigid body) -->
  <joint name="left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel"/>
    <!-- lift wheel by its radius, rotate 90Â° (this is in radians, and 1.57 radians ~ 90deg or pi/2) around Y to align cylinder axis -->
    <origin xyz="${-max_wheel_movement_down} ${wheel_y_offset} ${wheel_z}" rpy="1.5708 0 0"/>
    <axis   xyz="0 0 1"/>
  </joint>
  <!-- Physical rigid body -->
  <link name="left_wheel">
    <visual>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <origin xyz="0 0 ${wheel_z}" rpy="1.5708 0 0"/>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
      </geometry>
    </collision>
    <xacro:inertial_cylinder mass="0.15"
                             length="${wheel_width}"
                             radius="${wheel_radius}">
      <origin rpy="1.5708 0 0"/>
    </xacro:inertial_cylinder>

  </link>

  <gazebo reference="left_wheel">
    <material>Gazebo/Blue</material>
  </gazebo>

  <!--=== RIGHT WHEEL ===-->
  <joint name="right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel"/>
    <origin xyz="${-max_wheel_movement_down} -${wheel_y_offset} ${wheel_z}" rpy="1.5708 0 0"/>
    <axis   xyz="0 0 1"/>
  </joint>
  <link name="right_wheel">
    <visual>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <origin xyz="0 0 ${wheel_z}" rpy="1.5708 0 0"/>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
      </geometry>
    </collision>
    <xacro:inertial_cylinder mass="0.15"
                             length="${wheel_width}"
                             radius="${wheel_radius}">
      <origin rpy="1.5708 0 0"/>
    </xacro:inertial_cylinder>

  </link>

  <gazebo reference="right_wheel">
    <material>Gazebo/Blue</material>
  </gazebo>


  <joint name="right_wheel_joint_front" type="fixed">
    <parent link="base_link"/>
    <child link="right_wheel_front"/>
    <origin xyz="${wheel_y_offset} 0 ${wheel_z}" rpy="1.5708 0 0"/>
    <axis   xyz="0 0 1"/>
  </joint>
  <link name="right_wheel_front">
    <visual>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin xyz="${wheel_y_offset} 0 ${wheel_z}" rpy="1.5708 0 0"/>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
      </geometry>
    </collision>
    <xacro:inertial_cylinder mass="0.15"
                             length="${wheel_width}"
                             radius="${wheel_radius}">
      <origin rpy="1.5708 0 0"/>
    </xacro:inertial_cylinder>

  </link>

  <gazebo reference="right_wheel_front">
    <material>Gazebo/TransparentOrange</material>
    <mu1 value="0.001"/>
    <mu2 value="0.001"/>
  </gazebo>


  <joint name="left_wheel_joint_front" type="fixed">
    <parent link="base_link"/>
    <child link="left_wheel_front"/>
    <origin xyz="${-wheel_y_offset} 0 ${wheel_z}" rpy="1.5708 0 0"/>
    <axis   xyz="0 0 1"/>
  </joint>
  <link name="left_wheel_front">
    <visual>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin xyz="${-wheel_y_offset} 0 ${wheel_z}" rpy="1.5708 0 0"/>
      <geometry>
        <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
      </geometry>
    </collision>
    <xacro:inertial_cylinder mass="0.15"
                             length="${wheel_width}"
                             radius="${wheel_radius}">
      <origin rpy="1.5708 0 0"/>
    </xacro:inertial_cylinder>

  </link>

  <gazebo reference="left_wheel_front">
    <material>Gazebo/TransparentOrange</material>
    <mu1 value="0.001"/>
    <mu2 value="0.001"/>
  </gazebo>

  <!--=== CASTER WHEEL ===-->
  <!-- <joint name="caster_wheel_joint" type="fixed">
    <parent link="base_link"/>
    <child link="caster_wheel"/>
    <origin xyz="${caster_x_offset} 0 ${-caster_radius/2}" rpy="0 0 0"/>
  </joint>
  <link name="caster_wheel">
    <visual>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
    </collision>
    <xacro:inertial_sphere
      mass="0.1"
      radius="${caster_radius}">
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </xacro:inertial_sphere>
  </link>

  <gazebo reference="caster_wheel">
    <material>Gazebo/White</material>
    <mu1 value="0.001"/>
    <mu2 value="0.001"/>
  </gazebo> -->

  <!-- <joint name="caster_wheel_joint_invisible" type="fixed">
    <parent link="base_link"/>
    <child link="caster_wheel_invisible"/>
    <origin xyz="${-caster_x_offset} 0 ${-caster_radius/2}" rpy="0 0 0"/>
  </joint>
  <link name="caster_wheel_invisible">
    <visual>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
    </collision>
    <xacro:inertial_sphere
      mass="0.1"
      radius="${caster_radius}">
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </xacro:inertial_sphere>
  </link>

  <gazebo reference="caster_wheel_invisible">
    <material>Gazebo/BlueTransparent</material>
    <mu1 value="0.001"/>
    <mu2 value="0.001"/>
  </gazebo> -->


  <!-- <joint name="caster_wheel_joint_invisible_l" type="fixed">
    <parent link="base_link"/>
    <child link="caster_wheel_invisible_l"/>
    <origin xyz="0 ${-caster_x_offset} ${-caster_radius/2}" rpy="0 0 0"/>
  </joint>
  <link name="caster_wheel_invisible_l">
    <visual>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
    </collision>
    <xacro:inertial_sphere
      mass="0.1"
      radius="${caster_radius}">
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </xacro:inertial_sphere>
  </link>

  <gazebo reference="caster_wheel_invisible_l">
    <material>Gazebo/BlueTransparent</material>
    <mu1 value="0.001"/>
    <mu2 value="0.001"/>
  </gazebo> -->


  <!-- <joint name="caster_wheel_joint_invisible_r" type="fixed">
    <parent link="base_link"/>
    <child link="caster_wheel_invisible_r"/>
    <origin xyz="0 ${caster_x_offset} ${-caster_radius/2}" rpy="0 0 0"/>
  </joint>
  <link name="caster_wheel_invisible_r">
    <visual>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
    </collision>
    <xacro:inertial_sphere
      mass="0.1"
      radius="${caster_radius}">
      <origin xyz="0 0 0" rpy="0 0 0"/>
    </xacro:inertial_sphere>
  </link>

  <gazebo reference="caster_wheel_invisible_r">
    <material>Gazebo/BlueTransparent</material>
    <mu1 value="0.001"/>
    <mu2 value="0.001"/>
  </gazebo> -->

  

  


</robot>